src/qt3utils/applications/controllers/andor_spectrometer.py

```diff
@@ -0,0 +1,524 @@
+    make_label_and_check_button,
+    make_label_frame,
+    make_tab_view,
+    make_popup_window_and_take_threaded_action,
+    prepare_list_for_option_menu,
+)
+
+
+class AndorSpectrometerController:
+
+    def __init__(self, logger_level: int):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logger_level)
+
+        self.spectrometer_config = andor.AndorSpectrometerConfig()
+
+        # Changing supported acquisition modes to remove "run until abort" which is nonsensical for this application.
+        self.spectrometer_config.SUPPORTED_ACQUISITION_MODES = (
+            self.spectrometer_config.AcquisitionMode.SINGLE_SCAN.name,
+            self.spectrometer_config.AcquisitionMode.ACCUMULATE.name,
+            self.spectrometer_config.AcquisitionMode.KINETICS.name,
+        )
+        self.spectrometer_daq = andor.AndorSpectrometerDataAcquisition(self.spectrometer_config)
+
+        self.spectrometer_config.open()
+
+        self.last_config_dict = {}
+
+        self.last_measured_spectrum = None
+        self.last_wavelength_array = None
+
+    @property
+    def clock_rate(self) -> float:
+        """
+        The clock rate of a single exposure (1/exposure_time in Hz).
+        """
+        exposure_time = self.spectrometer_config.exposure_time
+        return 1 / exposure_time if exposure_time > 0 else np.inf
+
+    def start(self) -> None:
+        """
+        Nothing to be done in this method. All acquisitions are happening in the "sample_spectrum" method.
+        """
+        self.logger.debug('Starting controller.')
+        self.spectrometer_daq.wait_for_target_temperature_if_necessary()
+
+    def stop(self) -> None:
+        """
+        Stopping data acquisition.
+
+        We do not abort the acquisition because when stop is
+        used, the scan will complete the row it is currently scanning.
+        We only need to abort acquisition when the controller closes.
+        """
+        self.logger.debug('Stopping controller.')
+        self.spectrometer_daq.stop_waiting_to_reach_temperature()
+
+    def close(self) -> None:
+        self.spectrometer_daq.close()
+        self.spectrometer_config.close()
+
+    def sample_spectrum(self) -> Tuple[np.ndarray, np.ndarray]:
+        self.logger.debug('Sampling Spectrum')
+        acq_mode = self.spectrometer_config.acquisition_mode
+        if acq_mode == self.spectrometer_config.AcquisitionMode.SINGLE_SCAN.name:
+            return self.spectrometer_daq.acquire('single')
+        elif acq_mode == self.spectrometer_config.AcquisitionMode.ACCUMULATE.name:
+            return self.spectrometer_daq.acquire('accumulation')
+        elif acq_mode == self.spectrometer_config.AcquisitionMode.KINETICS.name:
+            return self.spectrometer_daq.acquire('kinetic series')
+
+    def configure(self, config_dict: dict) -> None:
+        """
+        This method is used to configure the spectrometer with the provided settings.
+        """
```

Comment:
In these docstrings, can you explain what each input it/where it will come from so new users can troubleshoot easier?